### PR TITLE
Vinai/update dolt sql session variables

### PR DESCRIPTION
**Pull request by VinaiRachakonda**
This pr fixes a bug where system tables were not being properly updated with dolt_commit and dolt_reset.

### CODE REVIEW

File: bats/sql-commit.bats [hunk 1]
```diff
@@ -115,4 +111,56 @@ teardown() {
...
+
+@test "DOLT_COMMIT updates session variables and keeps dolt_diff fresh" {
+    run dolt sql << SQL
+SELECT DOLT_COMMIT('-a', '-m', 'Commit1');
+SELECT count(*) FROM dolt_diff_test WHERE to_commit = hashof('head');
+SQL
+
+    [ $status -eq 0 ]
+    # Represents that the diff table marks a change from the recent commit.
+    [[ "$output" =~ "1" ]] || false
```
Comment (zachmu) [inline bats/sql-commit.bats:134]: Doesn't dolt_commit() return  a hash? This could match it by accident

File: bats/sql-reset.bats [hunk 2]
```diff
@@ -172,3 +172,80 @@ teardown() {
      [[ "$output" =~ pk ]] || false
      [[ "$output" != 1  ]] || false
 }
+
+@test "DOLT_RESET --hard properly wipes diff system table immediately" {
```
Comment (zachmu) [inline bats/sql-reset.bats:176]: Wouldn't phrase this as wiping a system table

Maybe "no rows in dolt_diff table after reset --hard"

File: bats/sql-reset.bats [hunk 3]
```diff
@@ -172,3 +172,80 @@ teardown() {
...
+INSERT INTO test VALUES (1);
+SELECT DOLT_RESET('--hard');
+SELECT count(*) FROM dolt_diff_test;
+SQL
+    [ $status -eq 0 ]
+    # Represents that the diff table marks a change from the recent commit.
+    [[ "$output" =~ "0" ]] || false
+}
+
+@test "DOLT_RESET --hard properly wipes status table immediately" {
```
Comment (zachmu) [inline bats/sql-reset.bats:187]: Same comment

File: bats/sql-reset.bats [hunk 4]
```diff
@@ -172,3 +172,80 @@ teardown() {
...
+
+@test "DOLT_RESET --hard properly wipes diff system table immediately" {
+      run dolt sql << SQL
+INSERT INTO test VALUES (1);
+SELECT DOLT_RESET('--hard');
+SELECT count(*) FROM dolt_diff_test;
+SQL
+    [ $status -eq 0 ]
+    # Represents that the diff table marks a change from the recent commit.
+    [[ "$output" =~ "0" ]] || false
```
Comment (zachmu) [inline bats/sql-reset.bats:184]: What does DOLT_RESET return? Could that select match 0?

Comment (zachmu) [reply bats/sql-reset.bats:184]: Easy way to make sure this isn't happening is just to run the count in a separate dolt select invocation

File: bats/sql-reset.bats [hunk 5]
```diff
@@ -172,3 +172,80 @@ teardown() {
...
+      head_variable=@@dolt_repo_$$_head
+      head_hash=$(get_head_commit)
+      run dolt sql << SQL
+INSERT INTO test VALUES (1);
+SELECT DOLT_RESET('--hard');
+SELECT $head_variable;
+SQL
+    echo $output
+    [ $status -eq 0 ]
+    # Represents that the diff table marks a change from the recent commit.
```
Comment (zachmu) [inline bats/sql-reset.bats:207]: ?

File: bats/sql-reset.bats [hunk 6]
```diff
@@ -172,3 +172,80 @@ teardown() {
...
+    [[ "$output" =~ $head_hash ]] || false
+}
+
+@test "DOLT_RESET soft does not wipe status table" {
```
Comment (zachmu) [inline bats/sql-reset.bats:211]: Same comment about phrasing

File: bats/sql-reset.bats [hunk 7]
```diff
@@ -172,3 +172,80 @@ teardown() {
...
+        run dolt sql << SQL
+INSERT INTO test VALUES (1);
+SELECT DOLT_RESET('test');
+SELECT table_name FROM dolt_status;
+SQL
+    [ $status -eq 0 ]
+    [[ "$output" =~ "test" ]] || false
```
Comment (zachmu) [inline bats/sql-reset.bats:218]: What about checking to see that tables that were staged are now not staged?

File: go/libraries/doltcore/env/actions/reset.go [hunk 8]
```diff
@@ -102,23 +102,30 @@ func resetHardTables(ctx context.Context, dbData env.DbData, apr *argparser.ArgP
 	return newHead, nil
 }
 
-func ResetHardTables(ctx context.Context, dbData env.DbData, apr *argparser.ArgParseResults, workingRoot, stagedRoot, headRoot *doltdb.RootValue) error {
+func ResetHardTables(ctx context.Context, dbData env.DbData, apr *argparser.ArgParseResults, workingRoot, stagedRoot, headRoot *doltdb.RootValue) (string, error) {
```
Comment (zachmu) [inline go/libraries/doltcore/env/actions/reset.go:105]: Method doc, and describe return value

File: go/libraries/doltcore/sqle/dfunctions/dolt_commit.go [hunk 9]
```diff
@@ -204,3 +215,20 @@ func (d DoltCommitFunc) Resolved() bool {
 func (d DoltCommitFunc) Children() []sql.Expression {
 	return d.children
 }
+
+// setHeadAndWorkingSessionRoot takes in a ctx and the new head hashstring and updates the session head and working hashes.
+func setHeadAndWorkingSessionRoot(ctx *sql.Context, headHashStr string) error {
+	key := ctx.GetCurrentDatabase() + sqle.HeadKeySuffix
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/dfunctions/dolt_commit.go:221]: Isn't there a common method for this

File: go/libraries/doltcore/sqle/dfunctions/dolt_commit.go [hunk 10]
```diff
@@ -204,3 +215,20 @@ func (d DoltCommitFunc) Resolved() bool {
...
+	dsess := sqle.DSessFromSess(ctx.Session)
+	hashType := sql.MustCreateString(query.Type_TEXT, 32, sql.Collation_ascii_bin)
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/dfunctions/dolt_commit.go:223]: Const of file-level var for this

File: go/libraries/doltcore/sqle/dfunctions/dolt_commit.go [hunk 11]
```diff
@@ -204,3 +215,20 @@ func (d DoltCommitFunc) Resolved() bool {
...
+
+	return dsess.Set(ctx, key, hashType, headHashStr)
+}
+
+func setHeadSessionRoot(ctx *sql.Context, headHashStr string) error {
+	key := ctx.GetCurrentDatabase() + sqle.HeadKeySuffix
+	dsess := sqle.DSessFromSess(ctx.Session)
+	hashType := sql.MustCreateString(query.Type_TEXT, 32, sql.Collation_ascii_bin)
+
+	return dsess.Session.Set(ctx, key, hashType, headHashStr)
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/dfunctions/dolt_commit.go:233]: If you're going to do this, make this a method on the dolt_session object, not a backdoor like this.

Comment (zachmu) [reply go/libraries/doltcore/sqle/dfunctions/dolt_commit.go:233]: SetSessionVarDirectly

And give it a good method doc on why you might want to do that

**Review by zachmu — COMMENTED** [review]
Looks pretty good, needs a little work before checking in.

File: bats/sql-commit.bats (hunk 1 above)
Comment (VinaiRachakonda) [reply bats/sql-commit.bats:134]: Good catch.

File: go/libraries/doltcore/sqle/dfunctions/dolt_commit.go (hunk 9 above)
Comment (VinaiRachakonda) [reply go/libraries/doltcore/sqle/dfunctions/dolt_commit.go:221]: Not that ik of.

File: bats/sql-reset.bats [hunk 12]
```diff
@@ -134,41 +145,142 @@ teardown() {
...
+    dolt sql -q "SELECT DOLT_RESET('test');"
+    run dolt sql -q "SELECT * FROM dolt_status;"
 
-      # Add and unstage the table with a soft reset. Make sure the same data exists.
-     dolt add .
+    [ $status -eq 0 ]
+    [[ "$output" =~ "false" ]] || false
+}
+
+@test "DOLT_RESET soft maintains staged session variable" {
```
Comment (zachmu) [inline bats/sql-reset.bats:256]: this doesn't describe what the test does

File: bats/sql-reset.bats [hunk 13]
```diff
@@ -134,41 +145,142 @@ teardown() {
...
+    [[ "$output" =~ ([[:space:]]*new table:[[:space:]]*test2) ]] || false
+
+    dolt add .
+    dolt sql -q "SELECT DOLT_RESET('--hard');"
+    run dolt status
+
+    [ "$status" -eq 0 ]
+    [[ "$output" =~ "On branch master" ]] || false
+    [[ "$output" =~ "nothing to commit, working tree clean" ]] || false
+
```
Comment (zachmu) [inline bats/sql-reset.bats:207]: extra line

File: bats/sql-commit.bats [hunk 14]
```diff
@@ -115,4 +111,61 @@ teardown() {
...
+    [ $status -eq 0 ]
+    # Represents that the diff table marks a change from the recent commit.
+    [[ "$output" =~ $original_hash ]] || false
+}
+
+@test "DOLT_COMMIT updates session variables" {
+    head_variable=@@dolt_repo_$$_head
+    run dolt sql << SQL
+SELECT DOLT_COMMIT('-a', '-m', 'Commit1');
+SELECT $head_variable = HASHOF('head');
```
Comment (zachmu) [inline bats/sql-commit.bats:142]: This isn't really good enough... you want to verify that this variable actually reflects the true state of the world, not just that it is internally consistent within a session. You should at least verify that the variable changed before and after the commit command.

**Review by zachmu — APPROVED** [review]
LG, just a couple small changes

**Pull request merged**